src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,108 @@
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getFirstName)
+                .sorted(Comparator.reverseOrder())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted(Comparator.comparing(User::getAge, Comparator.reverseOrder())
+                        .thenComparing(User::getFirstName))
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .flatMap(user -> user.getPrivileges().stream())
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getAge() > age && user.getPrivileges().contains(Privilege.UPDATE))
+                .findAny();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors.groupingBy(user -> user.getPrivileges().size()));
     }
 
+
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToInt(User::getAge)
+                .average()
+                .orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+       return users.stream()
+               .map(User::getLastName)
+               .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
+               .entrySet().stream()
+               .filter(entry -> entry.getValue() > 1)
+               .collect(Collectors.groupingBy(Map.Entry::getValue))
+               .entrySet().stream()
+               .max(Map.Entry.comparingByKey())
+               .filter(entry -> entry.getValue().size() == 1)
+               .map(entry -> entry.getValue().get(0).getKey());
```

Comment:
Glad that you have 2 filters, that cover multiple cases, but for this task, you can minimize it a little bit. 
```suggestion
       return users.stream()
                .collect(Collectors.groupingBy(User::getLastName, Collectors.counting()))
                .entrySet()
                .stream()
                .max(Map.Entry.comparingByValue())
                .filter(map -> map.getValue().intValue() > 1)
                .map(Map.Entry::getKey);
```

Reply:
Take it as advice for the future, for now, leave it as it is since your implementation is very-very similar.